File: code/boj/10816_binary.js
const filePath =
  process.platform === "linux"
    ? "/dev/stdin"
    : "/Users/jeonsiwon/Desktop/coding-interview/code/boj/input.txt";
const input = require("fs")
  .readFileSync(filePath, "utf-8")
  .toString()
  .split("\n");

arr = input[1]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);
targetNum = input[3].split(" ").map(Number);

function lowerBound(target) {
  let start = 0;
  let end = arr.length - 1;
  while (start <= end) {
    const mid = Math.floor((start + end) / 2);
    if (arr[mid] < target) start = mid + 1;
    else end = mid - 1;
  }
  return start;
}

function upperBound(target) {
  let start = 0;
  let end = arr.length - 1;
  while (start <= end) {
    const mid = Math.floor((start + end) / 2);
    if (arr[mid] <= target) start = mid + 1;
    else end = mid - 1;
  }
  return start;
}

function solution() {
  let ans = "";
  for (let i = 0; i < targetNum.length; i++) {
    ans += `${upperBound(targetNum[i]) - lowerBound(targetNum[i])} `;
  }
  console.log(ans.trim());
}
solution();

File: problem/boj/10816_binary.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N,M,num;
vector<int> src;

void input(){
  cin >> N;
  src.resize(N);
  for(int i=0;i<N;i++){
    cin >> num;
    src[i] = num;
  } 
  sort(src.begin(),src.end());
  cin >> M;
  for(int i=0;i<M;i++){
    cin >> num;
    cout << upper_bound(src.begin(), src.end(), num)- lower_bound(src.begin(), src.end(), num)  << " ";
  }
  
}

void solution(){
  input();
}

int main(void){
  ios::sync_with_stdio(false);
  cin.tie(0);
  cout.tie(0);
  freopen("input.txt","r", stdin);
  solution();
}